src/collector/mod.rs

```diff
@@ -1,253 +1,314 @@
+    fn for_segment(
+        &self,
         segment_local_id: SegmentLocalId,
         segment: &SegmentReader,
-    ) -> Result<()>;
-    /// The query pushes the scored document to the collector via this method.
-    fn collect(&mut self, doc: DocId, score: Score);
+    ) -> Result<Self::Child>;
 
     /// Returns true iff the collector requires to compute scores for documents.
     fn requires_scoring(&self) -> bool;
-}
 
-impl<'a, C: Collector> Collector for &'a mut C {
-    fn set_segment(
-        &mut self,
-        segment_local_id: SegmentLocalId,
-        segment: &SegmentReader,
-    ) -> Result<()> {
-        (*self).set_segment(segment_local_id, segment)
-    }
-    /// The query pushes the scored document to the collector via this method.
-    fn collect(&mut self, doc: DocId, score: Score) {
-        C::collect(self, doc, score)
+    /// Combines the fruit associated to the collection of each segments
+    /// into one fruit.
+    fn merge_fruits(&self, children: Vec<Self::Fruit>) -> Self::Fruit;
+
+    /// You should not use this method.
+    ///
+    /// Instead, please use [the `Searcher`'s search(...) method](../struct.Searcher.html#method.search).
+    #[doc(hidden)]
+    fn search(&self, searcher: &Searcher, query: &Query) -> Result<Self::Fruit> {
+        let scoring_enabled = self.requires_scoring();
+        let weight = query.weight(searcher, scoring_enabled)?;
+        let mut fruits = Vec::new();
+        for (segment_ord, segment_reader) in searcher.segment_readers().iter().enumerate() {
+            let mut child: Self::Child = self.for_segment(segment_ord as SegmentLocalId, segment_reader)?;
+            let mut scorer = weight.scorer(segment_reader)?;
+            let delete_bitset_opt = segment_reader.delete_bitset();
+            if let Some(delete_bitset) = delete_bitset_opt {
+                scorer.for_each(&mut |doc, score|
+                    if !delete_bitset.is_deleted(doc) {
+                        child.collect(doc, score);
+                    });
+                fruits.push(child.harvest());
+            } else {
+                scorer.for_each(&mut |doc, score| child.collect(doc, score));
+                fruits.push(child.harvest());
+            }
+
+        }
+        Ok(self.merge_fruits(fruits))
     }
 
-    fn requires_scoring(&self) -> bool {
-        C::requires_scoring(self)
+    /// You should not use this method.
+    ///
+    /// Instead, please use [the `Searcher`'s search(...) method](../struct.Searcher.html#method.search).
+    #[doc(hidden)]
+    fn search_multithreads(&self, searcher: &Searcher, query: &Query, num_threads: usize) -> Result<Self::Fruit> {
+        let segment_readers = searcher.segment_readers();
+        let actual_num_threads = (segment_readers.len() + 1).max(num_threads);
+        let scoring_enabled = self.requires_scoring();
+        let weight = query.weight(searcher, scoring_enabled)?;
+        let thread_pool = rayon::ThreadPoolBuilder::new()
```

Comment:
So this builds a new threadpool with each search? Is this a perf concern?

Reply:
Excellent point.